camel/retrievers/bm25_retriever.py

```diff
@@ -0,0 +1,135 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from typing import Any, Dict, List
+
+import numpy as np
+
+from camel.loaders import UnstructuredIO
+from camel.retrievers import BaseRetriever
+
+DEFAULT_TOP_K_RESULTS = 1
+
+
+class BM25Retriever(BaseRetriever):
+    r"""An implementation of the `BaseRetriever` using the `BM25` model.
+
+        This class facilitates the retriever of relevant information using a
+        query-based approach, it ranks documents based on the occurrence and
+        frequency of the query terms.
+
+        Attributes:
+            bm25 (BM25Okapi): An instance of the BM25Okapi class used for
+                calculating document scores.
+            content_input_path (str): The path to the content that has been
+                processed and stored.
+            chunks (List[Any]): A list of document chunks processed from the
+                input content.
+
+        References:
+            https://github.com/dorianbrown/rank_bm25
+    """
+
+    def __init__(self) -> None:
+        r"""Initializes the BM25Retriever.
+        """
+
+        try:
+            from rank_bm25 import BM25Okapi
+        except ImportError as e:
+            raise ImportError("Package `rank_bm25` not installed.") from e
+
+        self.bm25: BM25Okapi = None
+        self.content_input_path: str = ""
+        self.chunks: List[Any] = []
+
+    def process(self, content_input_path: str, **kwargs: Any) -> None:
+        r"""Processes content from a file or URL, divides it into chunks by
+        using `Unstructured IO`,then stored internally. This method must be
+        called before executing queries with the retriever.
+
+        Args:
+            content_input_path (str): File path or URL of the content to be
+                processed.
+            **kwargs (Any): Additional keyword arguments for content parsing.
+        """
+        from rank_bm25 import BM25Okapi
+
+        # Load and preprocess documents
+        self.content_input_path = content_input_path
+        unstructured_modules = UnstructuredIO()
+        elements = unstructured_modules.parse_file_or_url(
+            content_input_path, **kwargs)
+        self.chunks = unstructured_modules.chunk_elements(
+            chunk_type="chunk_by_title", elements=elements)
+
```

Comment:
here we used `unstructured io` module do the chunking, but currently this module only have `chunk_by_title ` one type, but you are right we can open the flexibility to users for now, so that when `unstructured io` supports more types we don't need to spend more effort to change the code